src/contrib/cache_req.rst

```diff
@@ -0,0 +1,290 @@
+		.attr_expiration = SYSDB_CACHE_EXPIRE,
+		.parse_name = true,
+		.ignore_default_domain = false,
+		.bypass_cache = false,
+		.only_one_result = true,
+		.search_all_domains = false,
+		.require_enumeration = false,
+		.allow_missing_fqn = false,
+		.allow_switch_to_upn = true,
+		.upn_equivalent = CACHE_REQ_USER_BY_UPN,
+		.get_next_domain_flags = SSS_GND_DESCEND,
+
+		.is_well_known_fn = NULL,
+		.prepare_domain_data_fn = cache_req_user_by_name_prepare_domain_data,
+		.create_debug_name_fn = cache_req_user_by_name_create_debug_name,
+		.global_ncache_add_fn = NULL,
+		.ncache_check_fn = cache_req_user_by_name_ncache_check,
+		.ncache_add_fn = cache_req_user_by_name_ncache_add,
+		.ncache_filter_fn = NULL,
+		.lookup_fn = cache_req_user_by_name_lookup,
+		.dp_send_fn = cache_req_user_by_name_dp_send,
+		.dp_recv_fn = cache_req_common_dp_recv,
+		.dp_get_domain_check_fn = NULL,
+		.dp_get_domain_send_fn = NULL,
+		.dp_get_domain_recv_fn = NULL,
+	};
+
+Next, ``cache_req_process_unit()`` is used to check if the input name contains a domain name
+, and if so it must match existing discovered domains by SSSD. If no domain is found SSSD
+falls back to using the UPN.
+
+SSSD then must choose a domain to search, in ``cache_req_select_domains()``. If a domain name
+was parsed above, SSSD can simply search that domain
+
+.. code-block:: sssd-log
+
+    [cache_req_select_domains] (0x0400): CR #11: Performing a single domain search
+
+Otherwise, SSSD will try to search each SSSD domain, one at a time.
+
+.. code-block:: sssd-log
+
+	[cache_req_select_domains] (0x0400): CR #10: Performing a multi-domain search
+
+.. note::
+    SSSD lookups may bypass the cache or the data provider. This determination is based on the ``cache_behavior`` enum which is also set at this time.
+    This can be seen in the logs here, `check` may be replaced with `bypass`.
+
+	*[cache_req_search_domains] (0x0400): CR #23: Search will check the cache and check the data provider*
+
+SSSD then needs to validate and prepare the intended domain to search. This is done in ``cache_req_search_domains_next()``.
+Now SSSD is ready to perform the search with this domain.
+
+.. code-block:: sssd-log
+
+    [cache_req_set_domain] (0x0400): CR #0: Using domain [implicit_files]
+
+Performing the search
+---------------------
+
+In a multi-domain search, the below logic repeats for each domain until a result is found, or SSSD searched all
+domains and no result was found. Let's assume the normal cache behavior(do not bypass).
+
+.. mermaid::
+
+    graph TD
+    1[Search start] --> A
+    A[Negative cache] --Not found--> B(Local cache)
+    B --Not found--> D{Data Provider}
+    B --Object found--> C[Expired?]
+    C --Object expired--> D
+    D --Object found in dp--> E[Update local cache]
+    D --Object not found --> F[Next domain]
+    F --> A
+    E --> G[Search local cache again]
+
+
+* First the search checks the negative cache(``cache_req_user_by_name_ncache_check()``)
```

Comment:
Graph mixed with bullet list and log snippets make this chapter content a bit hard to follow.